Create Hector input .ini files programatically

mnichol3 opened the pull request:
This PR introduces the ability to create Hector input .ini files for a given scenario.

The file `R/create_ini.R` holds the function `create_scenario_ini` which writes an input .ini file for a given scenario string. This is accomplished by replacing placeholder strings in the template .ini file `data-raw/hector_template.ini` with the literals needed to create a Hector-readable input file. The created input file is written to the `input/` directory.

Closes #2 , closes #4 , closes #5

mnichol3 committed:
Initial commit

mnichol3 committed:
Updated for renv

mnichol3 committed:
Initial commit

mnichol3 committed:
Remove ini library import

mnichol3 committed:
Initial commit; Hector template .ini file

mnichol3 committed:
Hardcode start & end dates, spinup bool, spinup duration

mnichol3 committed:
Remove template ini param

mnichol3 committed:
Initial commit

mnichol3 committed:
Initial commit

mnichol3 committed:
Add documentation

mnichol3 committed:
Rename parse_template.R as parse_template_data.R

mnichol3 committed:
Update title, desc, authors, & license

mnichol3 committed:
Add documentation

mnichol3 committed:
Initial commit; Set up tests & test parse_emission_fname

mnichol3 committed:
Move test file to tests/testthat/

mnichol3 committed:
Syntax errors

mnichol3 committed:
Syntax fix

mnichol3 committed:
Add more test cases for parse_ini_template.R

* Test parse_ini_fname returns proper Hector .ini filename
* Test create_scenario_ini RCP45 .ini file; compare to default RCP_45
    .ini file from Hector package
* Test create_scenario_ini SSP245 .ini file; compare to default SSP245
    .ini file from Hector package

mnichol3 committed:
Add function to parse .ini file name

mnichol3 committed:
Documentation and whitespace formatting

mnichol3 committed:
Initial commit

mnichol3 committed:
Syntax fixes

* Resolve multiple-maintainer error in DESCRIPTION (sorry Kalyn)
* Fix oopsy-woopsy in parse_ini_fname test

mnichol3 committed:
Fix file reading synax

mnichol3 committed:
Initial commit

mnichol3 committed:
Use roxygen to generate docs

mnichol3 committed:
Solve internal package relative paths

mnichol3 committed:
Update roxygen version

mnichol3 committed:
Finally got this dumpster file working

mnichol3 committed:
Synax fixes UwU

mnichol3 committed:
Rename file to match ini filename

mnichol3 committed:
Initial commit

mnichol3 committed:
Add prefix to var_emissionsPath placeholder

* Prepend "emissions/" to "var_emissionsPath" placeholder string
  to match Hector .ini convention

mnichol3 committed:
Update to reflect changes to template .ini file

mnichol3 committed:
Work out kinks

* Fix bug that returned empty string as path of Hector .ini files from
  installed Hector package.
* Subsequent to above fix, fix bug that would erase contents of .ini
  files in installed Hector package.
* Break ini creation function into two helper functions to make life
  easier.
* Add tests to compare ini file created from template with ini file
  from installed Hector package.
* Add helper function(s) for testing ini creation function.

mnichol3 committed:
Rename "parse_ini_template.R" as "create_ini.R"

mnichol3 committed:
Rebuild

mnichol3 committed:
Remove old code, add doc

mnichol3 committed:
Remove file

mnichol3 committed:
Add export & keyword roxygen tags where appropriate

mnichol3 committed:
Initial commit; generated via roxygen

mnichol3 committed:
Initial commit; roxygen documentation

mnichol3 committed:
Remove unused file

mnichol3 committed:
Tweak docstring

mnichol3 committed:
Add function to check for scenario emissions file

kdorheim commented on .Rbuildignore:
```diff
@@ -1,2 +1,4 @@
+^renv$
+^renv\.lock$
 ^.*\.Rproj$
 ^\.Rproj\.user$
```
When you run`devtools::check('hectordata')` do you see a message like? 

```
N  checking top-level files
   Non-standard file/directory found at top level:
     ‘data-raw’
```

This is because the `data-raw` directory is missing from the .Rbuildignore file. If you created data-raw with `devtools::use_data_raw()` which I think has been deprecated and replaced by `usethis::use_data_raw` would have added it to this file automatically.

kdorheim commented on DESCRIPTION:
```diff
@@ -1,11 +1,15 @@
+LazyData: true
+Suggests:
+    testthat
+RoxygenNote: 6.1.1
```
According to package check `hectordata` dependence on R (>= 2.10), that should be added to this file. See the example here https://cran.r-project.org/doc/manuals/R-exts.html#The-DESCRIPTION-file for how to add this.

kdorheim commented on R/create_ini.R:
```diff
@@ -0,0 +1,126 @@
+#'
+#' @export
+create_scenario_ini <- function(scenario, check_em = TRUE) {
+  input_dir <- file.path(system.file('input', package='hectordata'))
```
Hmm... if this is a function that we are going to export I am not sure if we want to force users to write to the `system.file('input', package = 'hectordata')` because that means a user could try to add ini files to a package that has already been compiled. Using `system.file` calls is great for accessing package data but is not the best way to write data. May be the `input_dir` could be added as an argument and to write data out this is where using `here::here` can be helpful.

kdorheim commented on data-raw/hector_template.ini:
```diff
@@ -0,0 +1,388 @@
+; Template for a Hector config file
```
Where did this ini file come from? Could you add a read me to detail where the ini file came from?

mnichol3 replied:
Am I allowed to put a readme in `data-raw/` or will the cause an issue when building the package?

kdorheim replied:
You can add any file to the data-raw the directory needs to be added to the Rbuild ignore file so all the contents of the dir will be ignored. 

Problems happen when building the package if there are files/directories in the root dir that are unexpected. There can be problems with the subdirectories and names of files end up being too long but that is pretty rare.

kdorheim commented on hectordata.Rproj:
```diff
@@ -18,3 +18,4 @@ StripTrailingWhitespace: Yes
 BuildType: Package
 PackageUseDevtools: Yes
 PackageInstallArgs: --no-multiarch --with-keep.source
+PackageRoxygenize: rd,collate,namespace
```
We don't want the .Rproj to take over the commit history, let's make sure that .Rproj and .RData are not being committed to the repo more than absolutely necessary. See Tools > Project Options

kdorheim commented on renv.lock:
```diff
@@ -0,0 +1,657 @@
+{
+  "R": {
```
I had to install `devtools` and `testthat` to get hectordata to work, do they need to be added to the renv.lock?

mnichol3 replied:
Both [devtools](https://github.com/kdorheim/hectordata/pull/7/files/7edf78cf1641601f004c09e99f07f216a751ddd0#diff-8e91a611b7e51b93373e838ac754ed59R166) and [testthat](https://github.com/kdorheim/hectordata/pull/7/files/7edf78cf1641601f004c09e99f07f216a751ddd0#diff-8e91a611b7e51b93373e838ac754ed59R565) are already in `renv.lock`. I'll try a fresh reload to see if they're being installed.

kdorheim commented on tests/testthat/helper_create_ini.R:
```diff
@@ -0,0 +1,20 @@
+
+fetch_hector_ini <- function(scenario) {
+  hector_ini_name <- parse_ini_fname(scenario)
+  hector_ini_path <- system.file('input', hector_ini_name, package='hector')
```
The tests were failing for me for awhile before I realized that I needed to install hector. Please add some sort of check to assert that the hector_ini_path is a file that exists.  

You are more familiar with `renv` than me but I was under the impression that  renv could be used to load all of these dependencies. Would that be from running `renv::restore` or `renv::load`?

mnichol3 replied:
`renv::restore()` (or `Rscript -e 'renv::restore()'`) will begin the installation of packages described in `renv.lock`. More on reproducing `renv` libraries [here](https://rstudio.github.io/renv/articles/renv.html#reproducibility)

kdorheim replied:
Okay when I try to run renv::restore() I run into some issues related to `nlme`

```
- 'ld: library not found for -lgfortran'
Error: install of package 'nlme' failed
```

Where are we using `nlme`?

mnichol3 replied:
Great question....we're not. I think it was installed as a dependency's dependency. I re-initialized the renv library and installed the dependencies via `install.packages(c("devtools", "usethis", "testthat"), dependencies=c("Depends", "Imports"))`, but `nlme` was installed again

kdorheim replied:
Hmmm interesting if you check out the reverse dependencies on CRAN https://cran.r-project.org/web/packages/nlme/index.html are we using any of those packages?  Usually I am not that picky about package dependencies so long as it is clear what needs to be installed. If we have to use `nlme` fine but I am not excited about it because it looks like the installation requires multiple steps.

kdorheim replied:
An alternative to reading the ini file from the installed Hector would be to pull from the remote branch I've done this with csv files for another project so something like... 
```  readr::read_csv(url("https://raw.githubusercontent.com/JGCRI/CMIP6/master/cmip6_archive_index.csv"))
```

kdorheim replied:
readr::read_csv(url("https://raw.githubusercontent.com/JGCRI/CMIP6/master/cmip6_archive_index.csv"))

kdorheim reviewed (changes requested):
Hi @mnichol3 great job! There are just a handful of comments that need to be addressed. 

1. We want to make sure that the package passes `devtools::test()` and `devtools::test()`
2. If we are going to export functions that write output places I think we need to let users define where the output will be written/how that path is determined. 
3. Let's make sure that renv is set up with all the required packages and does not include dependencies we don't use.

mnichol3 committed:
First batch of PR suggestions from @kdorheim

* Re-initialize data-raw/ using usethis::use_data_raw
* Clean & rebuild renv library
  * Install devtools, usethis, & testthat, along with documented Depends
    and Imports
* Remove roxygen examples tags from create_ini.R docstrings

mnichol3 committed:
Remove packages from global lib

mnichol3 committed:
Second batch of PR changes

* Allow user to specify .ini output directory
  * check_emissions_file raises warning instead of error if
    corresponding EM file is not found in the directory
* Use Hector .ini file (tests/testthat/hector_test_rcp45.ini)
  for testing created .ini files instead of attempting to retrieve
  .ini files from Hector package (which may or may not be installed)
* Other stuff

mnichol3 committed:
Create data-raw readme

mnichol3 committed:
Formatting

mnichol3 committed:
Merge branch 'dev-ini_template' of https://github.com/mnichol3/hectordata into dev-ini_template

mnichol3 committed:
Fix default output path

mnichol3 committed:
Documentation typo

kdorheim reviewed (approved):
@mnichol3 great job! Thanks for all your hard work on this I am going to go ahead squash the commits and merge your PR
